ouroboros-consensus/test-storage/Test/Ouroboros/Storage/ChainDB/Iterator.hs

```diff
@@ -0,0 +1,301 @@
+a, b, c, d, e :: TestBlock
+a = mkBlk [0]
+b = mkBlk [0,0]
+c = mkBlk [0,0,0]
+d = mkBlk [0,0,0,0]
+e = mkBlk [0,0,0,0,0]
+
+-- | Requested stream = A -> C
+
+--              ImmDB              VolDB
+-- Hash    A -> B -> C -> D        C, D
+--
+-- Bug: we find a partial path [B]->C in the VolDB. Now the ForkTooOld
+-- condition is triggered because the tip of the ImmDB is not B but D.
+--
+-- For more details, see:
+-- https://github.com/input-output-hk/ouroboros-network/pull/773#issuecomment-513128004
+prop_773_bug :: Property
+prop_773_bug = prop_general_test
+    TestSetup
+      { immutable = Chain.fromOldestFirst [a, b, c, d]
+      , volatile  = [c, d]
+      }
+    (StreamFromInclusive (blockPoint a))
+    (StreamToInclusive   (blockPoint c))
+    (Right (map Right [a, b, c]))
+
+-- | Requested stream = A -> E
+
+--              ImmDB              VolDB
+-- Hash    A -> B -> C -> D        C   D   E
+--
+-- This was/is handled correctly in @streamFromBoth@.
+prop_773_working :: Property
+prop_773_working = prop_general_test
+    TestSetup
+      { immutable = Chain.fromOldestFirst [a, b, c, d]
+      , volatile  = [c, d, e]
+      }
+    (StreamFromInclusive (blockPoint a))
+    (StreamToInclusive   (blockPoint e))
+    (Right (map Right [a, b, c, d, e]))
+
+-- | The general property test
+prop_general_test
+  :: TestSetup
+  -> StreamFrom TestBlock
+  -> StreamTo   TestBlock
+  -> IterRes
+  -> Property
+prop_general_test setup from to expected =
+    counterexample (testSetupInfo setup) $
+    case (actual, expected) of
+      (Left actualErr, Left expectedErr)         -> actualErr === expectedErr
+      (Left actualErr, Right expectedStream)     -> failure $
+        "Got " <> show actualErr <> "\nbut expected " <> ppStream expectedStream
+      (Right actualStream, Left expectedErr)     -> failure $
+        "Got " <> ppStream actualStream <> "\nbut expected " <> show expectedErr
+      (Right actualStream, Right expectedStream)
+        | actualStream == expectedStream
+        -> property True
+        | otherwise
+        -> failure $ "Got " <> ppStream actualStream <> "\nbut expected " <>
+            ppStream expectedStream
+  where
+    (_trace, actual) = runIterator setup from to
+    failure msg = counterexample msg False
+
+    ppStream :: [Either (HeaderHash TestBlock) TestBlock] -> String
+    ppStream = intercalate " :> " . map ppEBBOrBlock
+
+{-------------------------------------------------------------------------------
+  Test setup
+-------------------------------------------------------------------------------}
+
+-- | The initial contents of the ImmutableDB and the VolatileDB.
+--
+-- Note that the iterator implementation does not rely on the current
+-- in-memory chain.
+data TestSetup = TestSetup
+  { immutable :: Chain TestBlock
+  , volatile  :: [TestBlock]
+  }
+
+mkBlk :: [Word64] -> TestBlock
+mkBlk h = TestBlock
+    { tbHash  = mkTestHash h
+    , tbSlot  = SlotNo $ fromIntegral $ 2 * length h
```

Comment:
How relevant is `2 * length h`? For example, does that merely make it slightly more interesting that 1 slot per block?

Reply:
Yeah, that's just so that we don't (somehow) rely on 1 slot per block. Ideally, we'd test with differently sized gaps, but the slot should be predictable since we want that `forall b1 b2. tbHash b1 == tbHash b2 => tbSlot b1 == tbSlot b2`.